sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashJoin.scala

```diff
@@ -126,11 +126,11 @@ trait HashJoin {
           while (currentHashMatches == null && streamIter.hasNext) {
             currentStreamedRow = streamIter.next()
             val key = joinKeys(currentStreamedRow)
-            if (!key.anyNull) {
-              currentHashMatches = hashedRelation.get(key)
-              if (currentHashMatches != null) {
-                currentMatchPosition = 0
-              }
+            // We do filtering null keys by inserting a Filter before inner join,
```

Comment:
The execution engines correctness should not be predicated on specific optimizations.  That is *really* tight coupling.  If you want to optimize code generation for null handling you should do it based on signals like `nullable` and not on undocumented, implicit contracts between components.